src/ardupilot/ArduPilotProtocol.cpp

```diff
@@ -0,0 +1,119 @@
+#include "ArduPilotProtocol.h"
+
+#include "../Constants.h"
+#include "../Globals.h"
+#include "../log.h"
+#include "../utils/json.h"
+
+#include <mutex>
+
+namespace net {
+namespace ardupilot {
+
+using val_t = json::value_t;
+using namespace robot::types;
+using namespace navtypes;
+using std::placeholders::_1;
+
+ArduPilotProtocol::ArduPilotProtocol() {}
+
+void ArduPilotProtocol::initArduPilotServer() {
+	auto ardupilot_protocol = std::make_unique<net::websocket::WebSocketProtocol>(
+		Constants::ARDUPILOT_PROTOCOL_NAME);
+	ardupilot_protocol->addMessageHandler(
+		"arduPilotGPSReport", std::bind(&ArduPilotProtocol::handleGPSRequest, this, _1),
+		validateGPSRequest);
+	ardupilot_protocol->addMessageHandler(
+		"arduPilotIMUReport", std::bind(&ArduPilotProtocol::handleIMURequest, this, _1),
+		validateIMURequest);
+	ardupilot_protocol->addMessageHandler(
+		"arduPilotHeadingReport",
+		std::bind(&ArduPilotProtocol::handleHeadingRequest, this, _1), validateHeadingRequest);
+	ardupilot_protocol->addConnectionHandler(
+		std::bind(&ArduPilotProtocol::clientConnected, this));
+	ardupilot_protocol->addDisconnectionHandler(
+		std::bind(&ArduPilotProtocol::clientDisconnected, this));
+
+	Globals::websocketServer.addProtocol(std::move(ardupilot_protocol));
+
+	{
+		log(LOG_INFO, "Waiting for ArduPilot to connect...\n");
+		std::unique_lock<std::mutex> lock(_connectionMutex);
+		_connectionCV.wait(lock, [this] { return _arduPilotProtocolConnected; });
+	}
```

Comment:
We probably don't want to block everything until the ardupilot connects. The ardupilot isn't strictly required for the rover to run, so we don't want to add that dependency.

When the ardupilot isn't connected we can just return empty datapoints.